### PR TITLE
C++ tensor multi-dim indexing: add Slice / TensorIndex

### PR DESCRIPTION
**Pull request by yf225**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30427 C++ tensor multi-dim indexing: more indexing tests
* #30426 Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h
* #30425 C++ tensor multi-dim indexing: add index() and index_put_() overloads, and simple indexing tests
* **#30424 C++ tensor multi-dim indexing: add Slice / TensorIndex**

`at::indexing::TensorIndex` is used for converting C++ tensor indices such as `{None, "...", Ellipsis, 0, true, {1, None, 2}, torch::tensor({1, 2})}` into its equivalent `std::vector<TensorIndex>`, so that further tensor indexing operations can be performed using the supplied indices.

Differential Revision: [D18695902](https://our.internmc.facebook.com/intern/diff/D18695902)

### CODE REVIEW

**Comment by kostmo** [comment]
## CircleCI build failures summary
As of commit 3a03b4f:
* **3/3** failures introduced in this PR



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=3a03b4f22e73579c88aa8e46eaf117000378170b):

### 3 failures recognized by patterns:
####  pytorch_linux_xenial_py2_7_9_test
**Step:** Test
```
Nov 25 22:48:52 [  FAILED  ] TensorIndexingTest.TensorIndex
```
####  pytorch_linux_xenial_cuda9_cudnn7_py3_multigpu_test
**Step:** Test
```
Nov 25 22:50:36 [  FAILED  ] TensorIndexingTest.TensorIndex
```
####  pytorch_linux_xenial_py3_6_gcc5_4_test
**Step:** Test
```
Nov 25 22:52:25 [  FAILED  ] TensorIndexingTest.TensorIndex
```


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

**Comment by ezyang** [comment]
Sorry, this diff fell off the cart. Looking at it now.

File: test/cpp/api/tensor_indexing.cpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    ASSERT_TRUE(indices[0].is_none());
+    ASSERT_TRUE(indices[1].is_ellipsis());
+    ASSERT_TRUE(indices[2].is_ellipsis());
+    ASSERT_TRUE(indices[3].is_integer());
+    ASSERT_TRUE(indices[3].integer() == 0);
+    ASSERT_TRUE(indices[4].is_boolean());
+    ASSERT_TRUE(indices[4].boolean() == true);
+    ASSERT_TRUE(indices[5].is_slice());
+    ASSERT_TRUE(indices[5].slice().start() == 1);
+    ASSERT_TRUE(indices[5].slice().stop() == INDEX_MAX);
```
Comment (ezyang) [inline test/cpp/api/tensor_indexing.cpp]: Interesting. You wrote `None` in the slice specification, but it turns into `INDEX_MAX`?  I guess this is sound because you can never out of bound slice ranges in PYthon.

File: test/cpp/api/tensor_indexing.cpp [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    ASSERT_TRUE(torch::equal(indices[6].tensor(), torch::tensor({1, 2})));
+  }
+
+  ASSERT_THROWS_WITH(
+    TensorIndex(".."),
+    "Expected \"...\" to represent an ellipsis index, but got \"..\"");
+
+  // NOTE: Some compilers such as Clang 5 and MSVC always treat `TensorIndex({1})` the same as
+  // `TensorIndex(1)`. Therefore, we should only run this test on compilers that treat
+  // `TensorIndex({1})` as `TensorIndex(std::initializer_list<c10::optional<int64_t>>({1}))`.
```
Comment (ezyang) [inline test/cpp/api/tensor_indexing.cpp]: What does the C++ standard say should happen in this case?

File: aten/src/ATen/native/TensorIndexing.h [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+struct CAFFE2_API EllipsisIndexType { EllipsisIndexType() {} };
+CAFFE2_API extern const EllipsisIndexType Ellipsis;
+
+struct CAFFE2_API Slice {
+ public:
+  Slice();
+  Slice(int64_t start, int64_t stop, int64_t step);
+
+  const int64_t& start() const;
```
Comment (ezyang) [inline aten/src/ATen/native/TensorIndexing.h]: Why do these return const references?

File: aten/src/ATen/native/TensorIndexing.h [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  // Case 1: `at::indexing::None`
+  TensorIndex(c10::nullopt_t);
+
+  // Case 2: "..." / `at::indexing::Ellipsis`
+  TensorIndex(at::indexing::EllipsisIndexType);
+  TensorIndex(const char *str);
+
+  // Case 3: Integer value
+  TensorIndex(int64_t integer);
+  TensorIndex(int integer);
```
Comment (ezyang) [inline aten/src/ATen/native/TensorIndexing.h]: Why do you also need an `int` overload here?

File: aten/src/ATen/native/TensorIndexing.h [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+  // Case 4: Boolean value
+  TensorIndex(bool boolean);
```
Comment (ezyang) [inline aten/src/ATen/native/TensorIndexing.h]: It might be a good idea to make sure that you don't do implicit conversions into boolean, since this can lead to some nasty bugs; e.g., your current constructor accepts a `std::shared_ptr<Tensor>`. You can do that with a template. (cc @zdevito)

File: aten/src/ATen/native/TensorIndexing.cpp [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+const int64_t& Slice::stop() const {
+  return stop_;
+}
+
+const int64_t& Slice::step() const {
+  return step_;
+}
+
+std::ostream& operator<<(std::ostream& stream, const Slice& slice) {
+  stream << "{" << slice.start() << ", " << slice.stop() << ", " << slice.step() << "}";
```
Comment (ezyang) [inline aten/src/ATen/native/TensorIndexing.cpp]: I am wondering why you stringify slices as the initializer list, as opposed to the Python syntax (e.g., `2::3` versus `{2, None, 3}`. It seems like it would be helpful to be able to report a slice in the Python syntax.

File: aten/src/ATen/native/TensorIndexing.cpp [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    c10::optional<int64_t> step_index = at::indexing::None) {
+  int64_t start, stop, step;
+  if (!step_index.has_value()) {
+    step = 1;
+  } else {
+    step = step_index.value();
+    if (step == 0) {
+      TORCH_CHECK(false, "slice step cannot be zero");
+    }
+    // Here *step might be -INDEX_MAX-1; in this case we replace it
```
Comment (ezyang) [inline aten/src/ATen/native/TensorIndexing.cpp]: you mean step?

**Review by ezyang — APPROVED** [review]
Looks reasonable.

File: aten/src/ATen/native/TensorIndexing.h [hunk 8]
```diff
@@ -0,0 +1,110 @@
+#pragma once
+
+#include <c10/util/Optional.h>
+#include <ATen/core/TensorBody.h>
```
Comment (yf225) [inline aten/src/ATen/native/TensorIndexing.h]: TODO: have a top level comment on this header explaining the general strategy here.

File: aten/src/ATen/native/TensorIndexing.h [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+namespace indexing {
+
+const int64_t INDEX_MAX = std::numeric_limits<int64_t>::max();
+const int64_t INDEX_MIN = std::numeric_limits<int64_t>::min();
+
+enum class TensorIndexType { None, Ellipsis, Integer, Boolean, Slice, Tensor };
+
+constexpr c10::nullopt_t None{c10::nullopt_t::init()};
+
+struct CAFFE2_API EllipsisIndexType { EllipsisIndexType() {} };
```
Comment (yf225) [inline aten/src/ATen/native/TensorIndexing.h]: TODO: make this `final`

File: aten/src/ATen/native/TensorIndexing.h [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+CAFFE2_API extern const EllipsisIndexType Ellipsis;
+
+struct CAFFE2_API Slice {
```
Comment (yf225) [inline aten/src/ATen/native/TensorIndexing.h]: TODO: make this `final`

File: aten/src/ATen/native/TensorIndexing.h [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+// `1::`                   | `{1, None, None}`
+// `:3`                    | `{None, 3}`
+// `:3:`                   | `{None, 3, None}`
+// `::2`                   | `{None, None, 2}`
+// `1:3`                   | `{1, 3}`
+// `1::2`                  | `{1, None, 2}`
+// `:3:2`                  | `{None, 3, 2}`
+// `1:3:2`                 | `{1, 3, 2}`
+// `torch.tensor([1, 2])`) | `torch::tensor({1, 2})`
+struct CAFFE2_API TensorIndex {
```
Comment (yf225) [inline aten/src/ATen/native/TensorIndexing.h]: TODO: make this `final`

File: aten/src/ATen/native/TensorIndexing.h (hunk 4 above)
Comment (yf225) [reply aten/src/ATen/native/TensorIndexing.h]: We need the `TensorIndex(int)` overload because otherwise e.g. `TensorIndex(5)` would be ambiguous:
```
../test/cpp/api/tensor_indexing.cpp:58:85: error: conversion from ‘int’ to ‘at::indexing::TensorIndex’ is ambiguous
     std::vector<TensorIndex> indices = {None, "...", Ellipsis, 0, true, {1, None, 2}};
                                                                                     ^
In file included from ../test/cpp/api/tensor_indexing.cpp:5:0:
../aten/src/ATen/native/TensorIndexing.h:70:3: note: candidate: at::indexing::TensorIndex::TensorIndex(int64_t)
   TensorIndex(int64_t integer);
   ^
../aten/src/ATen/native/TensorIndexing.h:67:3: note: candidate: at::indexing::TensorIndex::TensorIndex(const char*)
   TensorIndex(const char *str);
   ^
```
